### PR TITLE
[4.0] Cleanup postinstall messages in update

### PR DESCRIPTION
**Pull request by chmst**
Pull Request for Issue #26109 .

### Summary of Changes
Remove unused Language strings form cpanel. Postinstall messaged are in the respective language files of the extensions. 
Remove old messages from the database during upgrade: hathor, recaptche-1 and joomla40, eaccelerator

Remove files
 administrator/components/com_admin/postinstall/htaccess.php 
 administrator/components/com_admin/postinstall/updatedefaultsettings.php

### Testing Instructions
Use the package of this PR for upgrading a 3.10 installation to j4. 

### Actual result BEFORE applying this Pull Request
Missing translations for some postinstallation messages 

### Expected result AFTER applying this Pull Request
Everything is translated in postinstallation messages. 

Old messages are deleted from the database, these are remaining in the database: 

PLG_TWOFACTORAUTH_TOTP_POSTINSTALL_TITLE
COM_CPANEL_WELCOME_BEGINNERS_TITLE
COM_CPANEL_MSG_STATS_COLLECTION_TITLE
PLG_SYSTEM_UPDATENOTIFICATION_POSTINSTALL_UPDATECACHETIME', 
COM_ACTIONLOGS_POSTINSTALL_TITLE
COM_PRIVACY_POSTINSTALL_TITLE
PLG_SYSTEM_HTTPHEADERS_POSTINSTALL_INTRODUCTION_TITLE'

### Documentation Changes Required
no

### CODE REVIEW

**Comment by brianteeman** [comment]
Looks like you are removing more strings than you need to

**Comment by richard67** [comment]
@chmst Your update SQL for postgresql is equal to the one for mysql, but it needs different names quoting.

File: administrator/components/com_admin/sql/updates/postgresql/4.0.0-2020-12-31.sql [hunk 1]
```diff
@@ -0,0 +1,3 @@
+DELETE FROM `#__postinstall_messages` WHERE `title_key` = 'TPL_HATHOR_MESSAGE_POSTINSTALL_TITLE';
+DELETE FROM `#__postinstall_messages` WHERE `title_key` = 'PLG_PLG_RECAPTCHA_VERSION_1_POSTINSTALL_TITLE';
+DELETE FROM `#__postinstall_messages` WHERE `title_key` = 'COM_CPANEL_MSG_JOOMLA40_PRE_CHECKS_TITLE';
```
Comment (richard67) [inline administrator/components/com_admin/sql/updates/postgresql/4.0.0-2020-12-31.sql]:
```suggestion
DELETE FROM "#__postinstall_messages" WHERE "title_key" = 'TPL_HATHOR_MESSAGE_POSTINSTALL_TITLE';
DELETE FROM "#__postinstall_messages" WHERE "title_key" = 'PLG_PLG_RECAPTCHA_VERSION_1_POSTINSTALL_TITLE';
DELETE FROM "#__postinstall_messages" WHERE "title_key" = 'COM_CPANEL_MSG_JOOMLA40_PRE_CHECKS_TITLE';
```
Use the right names quoting for PostgreSQL.

**Comment by chmst** [comment]
Thanks @richard67, 
@brianteeman - I have spent much time with testing in different constellations but could not find a non-translated string. So any better test is appreciated.

**Comment by adj9** [comment]
I have tested this item :white_check_mark: successfully on 5a00c30061213738e1096aaa816a7def627adf6a<br /><br />Done :)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31838">issues.joomla.org/tracker/joomla-cms/31838</a>.</sub>

**Comment by brianteeman** [comment]
in 3.9 updated sites got a message COM_CPANEL_MSG_ADDNOSNIFF_BODY which i think is triggered by https://github.com/joomla/joomla-cms/tree/4.0-dev/administrator/components/com_admin/postinstall

**Comment by brianteeman** [comment]
I think you need to review all the files in https://github.com/joomla/joomla-cms/tree/4.0-dev/administrator/components/com_admin/postinstall and not just the one I highlighted. It could be that some of those fiules can be removed as you have removed the text in this pr or that you have to put the files back.

Basically you have done more in this pr than you described in the original post

**Comment by wilsonge** [comment]
We need to remove `COM_CPANEL_MSG_EACCELERATOR_TITLE` too. I think this is good otherwise

**Comment by wilsonge** [comment]
Brian is kinda right too though. I think we can remove `admin_postinstall_updatedefaultsettings_condition` - which are 3.8.8 default settings changes. We can probably also remove the htaccess changes as well. I don't think we need to keep these forever. We can remind users to check outstanding postinstall messages as part of the upgrade docs.

**Comment by richard67** [comment]
~~@chmst You have to rename the 2 update SQL scripts `4.0.0-2020-12-31.sql` to a date later than the last Beta release, which was on February 2, eg. to `4.0.0-2021-02-28.sql`. Otherwise they will not run when updating a Beta 7 to the next Beta or RC.~~

Update: Done after talk on Glip.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on a2c06a38bae6eaf43cc8effa202d357e43c050b4<br /><br />I've done 3 tests on 2 identical copy of my homepage which I had updated to 3.10 Alpha 5 and cleaned up by a not 4.0 compatible extension:
1. Updated copy 1 to latest 4.0 nightly without this PR applied to reproduce the issue.
2. Updated copy 2 to the update package for this PR to make sure the PR works as expected when updating from 3.10.
3. Updated the result of test 1 to the update package for this PR to make sure the PR works as expected when updating from 4.0 Beta 7 or later nightly.

The PHP files are not removed with the update.

They have to be added to script.php to the list of files to be deleted before making the next 4.0 release of whichever kind (beta or RC). Normally @wilsonge cares for that.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31838">issues.joomla.org/tracker/joomla-cms/31838</a>.</sub>

**Comment by richard67** [comment]
@brianteeman Could you give this release blocker here a test? Thanks in advance, it would be really appreciated.

File: administrator/language/en-GB/com_cpanel.ini [hunk 2]
```diff
@@ -20,9 +20,9 @@ COM_CPANEL_MSG_ADDNOSNIFF_BODY="<p>Joomla is now shipped with additional securit
 COM_CPANEL_MSG_ADDNOSNIFF_TITLE=".htaccess & web.config Security Update"
 COM_CPANEL_MSG_HTACCESS_BODY="A change to the default .htaccess and web.config files was made in Joomla! 3.4 to disallow folder listings by default.  Users are recommended to implement this change in their files.  Please see <a href=\"https://docs.joomla.org/Special:MyLanguage/Preconfigured_htaccess\">this page</a> for more information."
```
Comment (Quy) [inline administrator/language/en-GB/com_cpanel.ini]: Delete?

Comment (Quy) [reply administrator/language/en-GB/com_cpanel.ini]: Also lines 33-34?

**Comment by brianteeman** [comment]
You are removing the file for the htaccess changes (as requested by @wilsonge) but the language strings are still present and the database record is not removed

Also you need to check the condition_file in the database and make sure that those files are removed if the record is removed as they will never be used

File: administrator/language/en-GB/com_cpanel.ini (hunk 2 above)
Comment (chmst) [reply administrator/language/en-GB/com_cpanel.ini]: Thanks!

**Comment by chmst** [comment]
Thank you. Yes, I have forgotten language strings, sorry. Condition files must be removed in the script (see description / testing instruction).

**Comment by brianteeman** [comment]
> ondition files must be removed in the script (see description / testing instruction).

I saw the comment about leaving that to @wilsonge and I dont agree that we should leave it to him (yes I am guilty of forgetting to add files to that list).

However that list is for how to handle files that exist on sites when they are upgraded but no longer present in the core - such as the two files you are removing here.

I am saying that there are more condition files that need to be removed from core

**Comment by richard67** [comment]
@brianteeman I've just checked the files used as action file or condition file by the removed postinstall messages. From my point of view all is ok regarding the sources:

- `administrator/components/com_admin/postinstall/eaccelerator.php`, `administrator/components/com_admin/postinstall/joomla40checks.php` and `plugins/captcha/recaptcha/postinstall/actions.php` have been removed already in the 4.0-dev sources
- `administrator/templates/hathor/postinstall/hathormessage.php` has also been removed from sources and should already be in script.php for removing the hathor template.

What remains is to check and if necessary complete file deletion in script.php.

@chmst To me it seems the messages with title keys `COM_CPANEL_MSG_HTACCESS_TITLE` and `COM_CPANEL_MSG_UPDATEDEFAULTSETTINGS_TITLE` should be deleted in database, too. They belong to those 2 PHP files deleted by this PR but currently they are not deleted with the update SQL script of this PR or any other, so they are still in database after an update.

Beside this, it would be better to use one delete statement with an `IN` clause and order the key in that clause alphabetically. Assuming my above assumption is right, this would be for MySQL:

```
DELETE FROM `#__postinstall_messages`
 WHERE `title_key`
    IN ('COM_CPANEL_MSG_EACCELERATOR_TITLE',
        'COM_CPANEL_MSG_HTACCESS_TITLE',
        'COM_CPANEL_MSG_JOOMLA40_PRE_CHECKS_TITLE',
        'COM_CPANEL_MSG_UPDATEDEFAULTSETTINGS_TITLE',
        'PLG_PLG_RECAPTCHA_VERSION_1_POSTINSTALL_TITLE',
        'TPL_HATHOR_MESSAGE_POSTINSTALL_TITLE');
```

For PostgreSQL use the same, just with different name quotes `"`.

**Comment by brianteeman** [comment]
> @brianteeman I've just checked the files used as action file or condition file by the removed postinstall messages. From my point of view all is ok regarding the sources:

Look at the Joomla 3 install sql and the postinstall messages table
https://github.com/joomla/joomla-cms/blob/cea0be4b0d02cfa26673599e5e52bb69be55b259/installation/sql/mysql/joomla.sql#L1669-L1678

Here you can find the location of all the postinstall files. They were not all in the component folder. Some of the messages that have now been removed had a postinstall file that also has to be removed and hasnt been.

For example we still have files such as https://github.com/joomla/joomla-cms/blob/4.0-dev/plugins/twofactorauth/totp/postinstall/actions.php

This file is completely redundant now

**Comment by richard67** [comment]
@brianteeman Well, I have checked only for those postinstall messages handled by this PR, I haven't checked the others.

**Comment by brianteeman** [comment]
As the PR is to "cleanup postinstall messages on update" then this PR is not complete

**Comment by richard67** [comment]
@brianteeman On a new installation of a clean 4.0-dev the postinstall message with title_key = 'PLG_TWOFACTORAUTH_TOTP_POSTINSTALL_TITLE' is still in database. Are you sure that this is redundant now?

**Comment by brianteeman** [comment]
My bad on that one - need more coffee. I swear when I checked it last night that message wasnt included any more. It probably should be removed but that is a different issue to this. Sorry for the confusion

**Comment by richard67** [comment]
> Sorry for the confusion

No problem, you made me check it again, and better one check more than one check missing.

File: administrator/components/com_admin/script.php [hunk 3]
```diff
@@ -522,7 +522,9 @@ public function deleteUnexistingFiles($dryRun = false, $suppressOutput = false)
 			'/administrator/components/com_admin/models/profile.php',
 			'/administrator/components/com_admin/models/sysinfo.php',
 			'/administrator/components/com_admin/postinstall/eaccelerator.php',
+			'/administrator/components/com_admin/postinstall/htaccess.php',
```
Comment (richard67) [inline administrator/components/com_admin/script.php]: You have added that to the section for Beta 1, see the comment here:
https://github.com/joomla/joomla-cms/blob/e3682eb40c71f3da3fa99dc31c8254bbe4f6fc0f/administrator/components/com_admin/script.php#L495
But I think it should go to the bottom of that array with a new comment for a "Beta 8" section.

@wilsonge What do you prefer? Or shall we lave the script.php thing to you?

File: administrator/components/com_admin/script.php [hunk 4]
```diff
@@ -522,7 +522,9 @@ public function deleteUnexistingFiles($dryRun = false, $suppressOutput = false)
...
 			'/administrator/components/com_admin/postinstall/joomla40checks.php',
+			'/administrator/components/com_admin/postinstall/updatedefaultsettings.php',
```
Comment (richard67) [inline administrator/components/com_admin/script.php]: See previous review comment.

Comment (chmst) [reply administrator/components/com_admin/script.php]: let @wilsonge decide - having all postinstall files together is easier

Comment (brianteeman) [reply administrator/components/com_admin/script.php]: They should be placed where they 're removed. Various peoples workflows would be impacted if they are listed as removed in the wrong version.

Comment (richard67) [reply administrator/components/com_admin/script.php]: @chmst See https://github.com/chmst/joomla-cms/pull/10 .

File: administrator/components/com_admin/script.php (hunk 3 above)
Comment (richard67) [reply administrator/components/com_admin/script.php]: See chmst#10 .

File: administrator/components/com_admin/sql/updates/postgresql/4.0.0-2021-02-28.sql [hunk 5]
```diff
@@ -0,0 +1,8 @@
+DELETE FROM "#__postinstall_messages"
+WHERE "title_key"
```
Comment (richard67) [inline administrator/components/com_admin/sql/updates/postgresql/4.0.0-2021-02-28.sql]:
```suggestion
 WHERE "title_key"
```

**Comment by richard67** [comment]
@chmst I was so free to solve a merge conflict in script.php because it was my just merged PR #32582 which caused it, and I committed also a small code style changes for one of the SQL scripts so it's same as the other one.

**Comment by chmst** [comment]
Thanks!

**Comment by richard67** [comment]
Will test again later or tomorrow if still necessary then.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on b01f59743f03c5cfebc085cda1d2ed1c7da5d042<br /><br />I've done 3 tests on 2 identical copy of my homepage which I had updated to 3.10 Alpha 5 and cleaned up by a not 4.0 compatible extension:

1. Updated copy 1 to latest 4.0 nightly without this PR applied to reproduce the issue.
2. Updated copy 2 to the update package for this PR to make sure the PR works as expected when updating from 3.10.
3. Updated the result of test 1 to the update package for this PR to make sure the PR works as expected when updating from 4.0 Beta 7 or later nightly.

In both cases 2 and 3 the PHP files and database records of the obsolete postinstall messages handled by this PR are removed with the update, and the remaining messages do not contain any untranslated language strings.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31838">issues.joomla.org/tracker/joomla-cms/31838</a>.</sub>

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on b01f59743f03c5cfebc085cda1d2ed1c7da5d042<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31838">issues.joomla.org/tracker/joomla-cms/31838</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31838">issues.joomla.org/tracker/joomla-cms/31838</a>.</sub>

**Pull request merged**